File: src/bytecode/Compiler.zig
const Compiler = @This();

const std = @import("std");
const Ast = @import("../parser/Ast.zig");
const Opcode = @import("./opcodes.zig").Opcode;

const InstructionStream = std.ArrayList(u8);

code: InstructionStream = .empty,
gpa: std.mem.Allocator,

ast: Ast,

pub fn init(gpa: std.mem.Allocator, ast: Ast) Compiler {
    return .{
        .gpa = gpa,
        .ast = ast,
    };
}

fn compile(self: *Compiler) !void {
    const roots = self.ast.nodes.get(0).data.range;

    for (roots.start..roots.end) |index| {
        const root = self.ast.nodes.get(index);
        try self.lowerRoot(root);
    }
}

fn lowerRoot(self: *Compiler, node: Ast.Node) !void {
    switch (node.tag) {
        .ret => {
            try self.lowerRet(node);
        },
        _ => unreachable,
    }
}

fn lowerRet(self: *Compiler, node: Ast.Node) !void {
    if (node.data.opt) |val_index| {
        try self.lowerExpr(val_index);
    }
    try self.code.append(self.gpa, Opcode.to(.ret));
}

fn lowerExpr(self: *Compiler, node: Ast.Node) !void {
    switch (node.tag) {
        .int => {
            try self.lowerInt(node);
        },
        _ => unreachable,
    }
}

fn lowerInt(self: *Compiler, node: Ast.Node) !void {
    try self.emit(Opcode.to(.load_imm_int));

    const value = self.ast.getSource(node);

    //TODO: handle parsing error.
    const int_value = std.fmt.parseInt(i64, value, 10) catch 0;
    const aw: std.Io.Writer.Allocating = .fromArrayList(self.gpa, &self.code);
    defer aw.deinit();
    try aw.writer.writeInt(i64, int_value, .little);
}

fn emit(self: *Compiler, byte: u8) !void {
    try self.code.append(self.gpa, byte);
}

File: bytecode/Compiler.cpp
#include "bytecode/Compiler.h"
#include <charconv>
#include <cstdint>
#include <string_view>
#include "bytecode/Opcode.h"
#include "parser/Ast.h"


namespace Nyx {
    void Compiler::compile() {
        const Node &rootNode = m_ast.node(0);
        const NodeRange roots = std::get<NodeRange>(rootNode.data);

        for (NodeIndex index = roots.start; index < roots.end; ++index) {
            lowerRoot(m_ast.node(index));
        }
    }

    void Compiler::lowerRoot(const Node &node) {
        switch (node.tag) {
            case NodeTag::Ret: {
                lowerRet(node);
            }
            default: {
            }
        }
    }

    void Compiler::lowerRet(const Node &node) {
        const auto index = node.index();
        if (index.has_value()) {
            const Node &expr = m_ast.node(index.value());
            lowerExpr(expr);
        }
        emit(static_cast<uint8_t>(Opcode::Ret));
    }

    void Compiler::lowerExpr(const Node &node) {
        switch (node.tag) {
            case NodeTag::Integer: {
                lowerInt(node);
                break;
            }
            default: {
            }
        }
    }


    void Compiler::lowerInt(const Node &node) {
        const auto int_str = m_ast.getSource(node);

        int64_t value = 0;
        // handle error
        std::from_chars(int_str.data(), int_str.data() + int_str.size(), value);

        emit(static_cast<uint8_t>(Opcode::LdaImmInt));

        for (uint8_t i = 0; i < 8; i++) {
            emit(static_cast<uint8_t>(value >> (8 * i) & 0xff));
        }
    }


    void Compiler::emit(uint8_t byte) { m_code.push_back(byte); }
} // namespace Nyx
